### DISCUSSION TITLE
Why are Degenerate Triangles Encountered in Surface Mesh?

### DISCUSSION THREAD

**Discussion by Connor56** [pyvista/pyvista#1941] [Q&A]
I am trying to create a 3D Delaunay Triangulation of the following points that have the shape below:

![cuboid_shaped_crystal](https://user-images.githubusercontent.com/34070858/145730055-72861762-0601-412a-8e0d-91445ba84a9b.png)

The points represent a crystal lattice in 3D space, I'm doing this to extract information about the surface of the lattice. So really what I want is the surface of the points, but finding it by delaunay3d and then extract_geometry is the only way I am currently aware of. Attempting this with the point set gives me the following:

![cuboid_shape_surface](https://user-images.githubusercontent.com/34070858/145730237-9ec97cec-3c8b-49fc-8d55-967d01a279b0.png)

As you can see from the picture there are points at which there is no surface, and I get the following warning when doing the delaunay3d triangulation:

> .2021-12-12 21:28:07.622 (  49.532s) [           8E149]      vtkDelaunay3D.cxx:522   WARN| vtkDelaunay3D (0x7fe02968acb0): 31 degenerate triangles encountered, mesh quality suspect

This makes it impossible for me to get the actual surface of the crystal structure because of error induced by these gaps. My pipeline is the following:

```
cloud = pv.PolyData(atom_array)
volume = cloud.delaunay_3d(alpha=2)
shell = volume.extract_geometry()
```

Changing the alpha does make a difference to the algorithm, but alpha's much larger than 2 don't produce the correct surface planes, and using delaunay3d without any alpha also doesn't produce the correct surface planes. I assume the problem is with the degenerate triangles, but I have no idea how to stop this being a problem.

Any help is greatly appreciated, on either Delaunay Triangulation, or another method for capturing the surface of the point system.

**Comment by akaszynski** [comment]
I'd try the [reconstruct_surface](https://docs.pyvista.org/examples/01-filter/surface_reconstruction.html) filter. It generally works better than `delaunay_3d`.

**Reply by Connor56** [reply]
Thankyou! Will it work with the PolyData object, or does it need to use something else? Also, following the link you have provided it seems that I need to have points that only represent the surface of the object? Is that correct, or can I used the same set of points that I am using for Delaunay3D?

**Reply by akaszynski** [reply]
The [pyvista.DataSetFilters.reconstruct_surface()](https://docs.pyvista.org/api/core/_autosummary/pyvista.DataSetFilters.reconstruct_surface.html#pyvista.DataSetFilters.reconstruct_surface) filter accepts PolyData containing points or points and faces. The following works just fine:

```py
import pyvista as pv
pv.Sphere().reconstruct_surface().plot()
```
If fact, it will return the same result as the [example](https://docs.pyvista.org/examples/01-filter/surface_reconstruction.html).

**Reply by Connor56** [reply]
Okay, but for an object that's point dense like mine above would this still work? As I currently understand it the `pv.Sphere` command doesn't produce a dense set of points, but points on the surface of a sphere.

**Reply by akaszynski** [reply]
A dense field or surface of points will still work. I'd try it and see if it works.

**Reply by Connor56** [reply]
Long time between replies, however, I have just tried this code again and got the following output:

`'PolyData' object has no attribute 'reconstruct_surface'`

I get the same from the example you have linked. Has an update changed how you access this filter?

**Reply by Connor56** [reply]
Looks like updating the pyvista package fixed my problem!

**Reply by Connor56** [reply]
Any idea how long this algorithm takes compared to delaunay trinagulation? I Just fed it about 300,000 points and it doesn't seem to be moving anywhere fast!

**Reply by banesullivan** [reply]
I'm not sure how long it might take with that dense point cloud, but you can try using the `progress_bar=True` argument to see the algorithm's progress

ref https://docs.pyvista.org/api/core/_autosummary/pyvista.PolyData.reconstruct_surface.html

**Reply by Connor56** [reply]
Yeah I tried, it was at 0% after 2000 seconds! I'm going to figure out another way to do it though!
